// express-react-views/index.js
/*
 *  Copyright (c) 2014, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 */

var React = require('react');
var ReactDOMServer = require('react-dom/server');
var beautifyHTML = require('js-beautify').html;
var assign = require('object-assign');

var DEFAULT_OPTIONS = {
  doctype: '<!DOCTYPE html>',
  beautify: false,
  transformViews: true,
};

function createEngine(engineOptions) {
  var registered = false;
  var moduleDetectRegEx;

  engineOptions = assign({}, DEFAULT_OPTIONS, engineOptions || {});

  function renderFile(filename, options, cb) {
    // Defer babel registration until the first request so we can grab the view path.
    console.log("--------------- react views options = \n filename + " + filename + " -- \n" + JSON.stringify(options,null, 2))
    if (!moduleDetectRegEx) {
      moduleDetectRegEx = new RegExp('^' + options.settings.views);
    }
    if (engineOptions.transformViews && !registered) {
      // Passing a RegExp to Babel results in an issue on Windows so we'll just
      // pass the view path.
      require('babel/register')({
        only: options.settings.views
      });
      registered = true;
    }

    try {
      var markup = engineOptions.doctype;
      console.log("----------loading in file:"+ filename)
      var component = require(filename);
      // Transpiled ES6 may export components as { default: Component }
      component = component.default || component;
      console.log("----------createElement:");
      var element = React.createElement(component, options);
      console.log("----------renderToStaticMarkup:"+ (JSON.stringify(element)));
      markup += ReactDOMServer.renderToStaticMarkup(element
      );
    } catch (e) {
      console.log("---- Got error " + JSON.stringify(e));
      return cb(e);
    }

    if (engineOptions.beautify) {
      // NOTE: This will screw up some things where whitespace is important, and be
      // subtly different than prod.
      console.log("----------beautify:"+ markup)
      markup = beautifyHTML(markup);
    }

    if (options.settings.env === 'development') {
      console.log("---------- deleting file from cache ")

      // Remove all files from the module cache that are in the view folder.
      Object.keys(require.cache).forEach(function(module) {
        if (moduleDetectRegEx.test(require.cache[module].filename)) {
          delete require.cache[module];
        }
      });
    }
    console.log("---------- DONE!! markup:"+ markup)

    cb(null, markup);
  }

  return renderFile;
}

exports.createEngine = createEngine;
